plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,472 @@
+	if _, ok := d.containerList[containerId]; ok {
+		return true
+	}
+	return false
+}
+
+func (d *DockerLogs) stopAllReaders() error {
+	for _, container := range d.containerList {
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		return errors.New(fmt.Sprintf("%s : Dock client is null", ERR_PREFIX))
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			logOptions := types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: true,
+				Timestamps: false,
+				Details:    true,
+				Follow:     true,
+				Tail:       "0",
+			}
+			logReader, err := d.client.ContainerLogs(context.Background(), c.ID, logOptions)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+			d.addToContainerList(c.ID, logReader)
+			err = d.tailContainerLogs(c, logReader, acc)
+			if err != nil {
+				acc.AddError(err)
+			}
+			d.removeFromContainerList(c.ID)
+			return
+		}(container)
+	}
+	return nil
+}
+
+func (d *DockerLogs) tailContainerLogs(
+	container types.Container, logReader io.ReadCloser,
+	acc telegraf.Accumulator,
+) error {
+	c, err := d.client.ContainerInspect(context.Background(), container.ID)
+	if err != nil {
+		return err
+	}
+	/* Parse container name */
+	var cname string
+	for _, name := range container.Names {
+		trimmedName := strings.TrimPrefix(name, "/")
+		match := d.containerFilter.Match(trimmedName)
+		if match {
+			cname = trimmedName
+			break
+		}
+	}
+
+	if cname == "" {
+		return errors.New(fmt.Sprintf("%s : container name is null", ERR_PREFIX))
+	}
+	imageName, imageVersion := parseImage(container.Image)
+	tags := map[string]string{
+		"container_name":    cname,
+		"container_image":   imageName,
+		"container_version": imageVersion,
+	}
+	fields := map[string]interface{}{}
+	fields["container_id"] = container.ID
+	// Add labels to tags
+	for k, label := range container.Labels {
+		if d.labelFilter.Match(k) {
+			tags[k] = label
+		}
+	}
+	if c.Config.Tty {
+		err = pushTtyLogs(acc, tags, fields, logReader)
+	} else {
+		_, err = pushLogs(acc, tags, fields, logReader)
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func pushTtyLogs(acc telegraf.Accumulator, tags map[string]string, fields map[string]interface{}, src io.Reader) (err error) {
+	tags["logType"] = "unknown" //in tty mode we wont be able to differentiate b/w stdout and stderr hence unknown
+	data := make([]byte, logBytesMax)
+	for {
+		num, err := src.Read(data)
+		if num > 0 {
+			fields["message"] = data[1:num]
```

Comment:
Why do we start at `1` instead of `data[0:num]`?  In my testing of long lines the first character is cut off:
```
docker_log,container_image=quay.io/influxdb/telegraf-ci,container_name=cocky_shirley,container_version=1.12.5,host=loaner,stream=stdout container_id="4a09148e7188863b4d4274f8049e1c21ebcef582f94fe670d131930de6c6fc12",message="ONTAINER_PARTIAL_MESSAGE=true ................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................" 1559630720000000000
```